### DISCUSSION TITLE
How to use PeasyCam (Java/.jar based Processing library)

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#110] [Q&A] [answered]
I have been meaning to try this for a looooong time, almost a year maybe? I've read this example many times: https://py5.ixora.io/how_tos/use_camera3D.html and now I've put the `peasycam.jar` and `peasy-math.jar` files in a subdirectory called `jars` next to my sketch source file.

Perhaps the `jpype.imports` is not need anymore, so I've tried with and without it, but I get this error:
```
    from PeasyCam import PeasyCam
ModuleNotFoundError: No module named 'PeasyCam'
```
I have also tried `from peasycam import PeasyCam` as it is `peasycam.jar` and the old  `add_library('peasycam')` went lowercase too. And fiddled with `py5tools.add_jars()`, but I think it shouldn't be necessary. 

Any tips?

Update: With `py5_tools.get_classpath()` I can see the peasycam.jar.
```
/home/villares/.../py5/jars/jogl-all.jar:
/home/villares/.../py5/jars/gluegen-rt.jar:
/home/villares/.../py5/jars/py5.jar:
/home/villares/.../py5/jars/core.jar:
/home/villares/.../py5/jars/pdf/pdf.jar:
/home/villares/.../py5/jars/pdf/itext.jar:
/home/villares/.../py5/jars/dxf/dxf.jar:
/home/villares/.../py5/jars/svg/svg.jar:
/home/villares/.../py5/jars/svg/batik.jar:
/home/villares/sketch/jars/peasy-math.jar:
/home/villares/sketch/jars/peasycam.jar
```

**Comment by hx2A** [comment, marked as answer]
> Update: With py5_tools.get_classpath() I can see the peasycam.jar.

That was a good debugging step. Sometimes you might think the jar is being included but it is actually not. I'm happy you thought to do that.

I also tried a bunch of import statements that didn't work. Eventually I stopped guessing and got the answer by looking at the [PeasyCam.java source code](https://github.com/jdf/peasycam/blob/master/src/peasy/PeasyCam.java). The package name for that class is `peasy`. The line of code you need is:

```python
from peasy import PeasyCam
```

There's also an [example notebook using PeasyCam in py5examples](https://github.com/py5coding/py5examples/blob/main/notebooks/PeasyCam%20Example.ipynb) with the correct import statement.

**Reply by villares** [reply]
It is working fine, I have the `jars` next to the sketch. 
I just have a question... should there be a way of pointing to a different `jars` location when on imported mode? Am I missing something?
I'm thinking about translating a series of examples and I would like to point to a single folder.
When on module mode one can do it before `import py5` but on imported mode it looks like a catch-22...

**Reply by hx2A** [reply]
You have a few options for ways to add jars before `import py5` gets called.

You can set the environment variable `$PY5_JARS` to a directory. Any jars in that directory or a subdirectory will be automatically added.

If you are using the run_sketch utility, you can add jars with the `--classpath` command line argument.

If you are using Thonny, the py5 plugin can be enhanced to support either of these two options if it does not do one of them already.

**Reply by villares** [reply]
Perfect! I'll try stuff and think about it. I think it would be nice to have a `jars` folder ready in the portable Thonny for people to install their Processing libraries and I believe the environment variable might be a good way of doing it.
